src/librbd/api/Pool.cc

```diff
@@ -0,0 +1,232 @@
+    if (m_scan_snaps) {
+      cls_client::get_snapcontext_start(&op);
+    }
+
+    m_out_bl.clear();
+    auto aio_comp = util::create_rados_callback<
+      C_ImageStat<I>, &C_ImageStat<I>::handle_get_head>(this);
+    int r = m_io_ctx.aio_operate(util::header_name(m_image_id), aio_comp, &op,
+                                 &m_out_bl);
+    ceph_assert(r == 0);
+    aio_comp->release();
+  }
+
+  void handle_get_head(int r) {
+    ldout(m_cct, 15) << "r=" << r << dendl;
+
+    auto it = m_out_bl.cbegin();
+    if (r == 0) {
+      uint8_t order;
+      r = cls_client::get_size_finish(&it, &m_max_size, &order);
+    }
+    if (m_scan_snaps && r == 0) {
+      r = cls_client::get_snapcontext_finish(&it, &m_snapc);
+      if (r == 0) {
+        (*m_snaps) += m_snapc.snaps.size();
+      }
+    }
+
+    if (r == -ENOENT) {
+      finish(r);
+      return;
+    } else if (r < 0) {
+      lderr(m_cct) << "failed to stat image: " << cpp_strerror(r) << dendl;
+      finish(r);
+      return;
+    }
+
+    if (!m_snapc.is_valid()) {
+      lderr(m_cct) << "snap context is invalid" << dendl;
+      finish(-EIO);
+      return;
+    }
+
+    get_snaps();
+  }
+
+  void get_snaps() {
+    if (!m_scan_snaps || m_snapc.snaps.empty()) {
+      finish(0);
+      return;
+    }
+
+    ldout(m_cct, 15) << dendl;
+    librados::ObjectReadOperation op;
+    for (auto snap_seq : m_snapc.snaps) {
+      cls_client::get_size_start(&op, snap_seq);
+    }
+
+    m_out_bl.clear();
+    auto aio_comp = util::create_rados_callback<
+      C_ImageStat<I>, &C_ImageStat<I>::handle_get_snaps>(this);
+    int r = m_io_ctx.aio_operate(util::header_name(m_image_id), aio_comp, &op,
+                                 &m_out_bl);
+    ceph_assert(r == 0);
+    aio_comp->release();
+  }
+
+  void handle_get_snaps(int r) {
+    ldout(m_cct, 15) << "r=" << r << dendl;
+
+    auto it = m_out_bl.cbegin();
+    for ([[maybe_unused]] auto snap_seq : m_snapc.snaps) {
+      uint64_t size;
+      if (r == 0) {
+        uint8_t order;
+        r = cls_client::get_size_finish(&it, &size, &order);
+      }
+      if (r == 0 && m_max_size < size) {
+        m_max_size = size;
+      }
+    }
+
+    if (r == -ENOENT) {
+      ldout(m_cct, 15) << "out-of-sync metadata" << dendl;
+      get_head();
+    } else if (r < 0) {
+      lderr(m_cct) << "failed to retrieve snap size: " << cpp_strerror(r)
+                   << dendl;
+      finish(r);
+    } else {
+      finish(0);
+    }
+  }
+
+};
+
+} // anonymous namespace
+
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::api::Pool: " << __func__ << ": "
+
+template <typename I>
+int Pool<I>::init(librados::IoCtx& io_ctx, bool force) {
+  auto cct = reinterpret_cast<CephContext*>(io_ctx.cct());
+  ldout(cct, 10) << dendl;
+
+  int r = io_ctx.application_enable(pg_pool_t::APPLICATION_NAME_RBD, force);
+  if (r < 0) {
+    return r;
+  }
+
+  // TODO configure self-managed snapshots (and other one-time pool checks)
```

Comment:
not sure about self-managed snapshots, as they are configured for data pool

Reply:
... you have to "rbd pool init" for any RBD pool to get the tags correct. It certainly doesn't hurt to also init the snapshot mode/verify EC override support at the same time. The ```CreateRequest``` state machine can then just re-use the same helper method.